controllers/persistent-volume-claim/controller.go

```diff
@@ -0,0 +1,102 @@
+
+import (
+	"context"
+	"github.com/openshift/lvm-operator/controllers"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/storage/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	crlog "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"strings"
+)
+
+// PersistentVolumeClaimReconciler reconciles a PersistentVolumeClaim object
+type PersistentVolumeClaimReconciler struct {
+	client    client.Client
+	apiReader client.Reader
+	recorder  record.EventRecorder
+}
+
+// NewPersistentVolumeClaimReconciler returns PersistentVolumeClaimReconciler.
+func NewPersistentVolumeClaimReconciler(client client.Client, apiReader client.Reader, eventRecorder record.EventRecorder) *PersistentVolumeClaimReconciler {
+	return &PersistentVolumeClaimReconciler{
+		client:    client,
+		apiReader: apiReader,
+		recorder:  eventRecorder,
+	}
+}
+
+//+kubebuilder:rbac:groups=core,resources=persistentvolumeclaims,verbs=get;list;watch;update
+//+kubebuilder:rbac:groups=storage.k8s.io,resources=csistoragecapacities,verbs=get;list;watch
+//+kubebuilder:rbac:groups=core,resources=events,verbs=create;update;patch
+
+// Reconcile PVC
+func (r *PersistentVolumeClaimReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := crlog.FromContext(ctx)
+
+	pvc := &corev1.PersistentVolumeClaim{}
+	err := r.client.Get(ctx, req.NamespacedName, pvc)
+	switch {
+	case err == nil:
+	case apierrors.IsNotFound(err):
+		return ctrl.Result{}, nil
+	default:
+		return ctrl.Result{}, err
+	}
+
+	// Skip if the PVC is deleted or does not use the lvms storage class.
+	if pvc.DeletionTimestamp != nil || !strings.HasPrefix(*pvc.Spec.StorageClassName, controllers.StorageClassPrefix) {
+		return ctrl.Result{}, nil
+	}
+
+	// Skip if the PVC is not in Pending state.
+	if pvc.Status.Phase != "Pending" {
+		return ctrl.Result{}, nil
+	}
+
+	// List csi storage capacities
+	storageCapacityList := &v1.CSIStorageCapacityList{}
+	err = r.client.List(ctx, storageCapacityList, &client.ListOptions{Namespace: controllers.OperatorNamespace})
+	switch {
+	case err == nil:
+	case apierrors.IsNotFound(err):
+		return ctrl.Result{}, nil
+	default:
+		return ctrl.Result{}, err
+	}
+
+	// Check if there is any node that has enough storage capacity for the PVC
+	found := false
+	for _, sc := range storageCapacityList.Items {
+		if strings.HasPrefix(sc.StorageClassName, controllers.StorageClassPrefix) && pvc.Spec.Resources.Requests.Storage().Cmp(*sc.Capacity) < 0 {
+			found = true
+		}
+	}
+
+	// Publish an event if the requested storage is greater than the available capacity
+	if !found {
+		r.recorder.Event(pvc, "Warning", "NotEnoughCapacity", "Requested storage is greater than available capacity on all the nodes. Check available capacity by `oc get csistoragecapacity`.")
```

Comment:
```suggestion
		r.recorder.Event(pvc, "Warning", "NotEnoughCapacity", "Requested storage is greater than available capacity on any node. Check available node capacity by `oc get csistoragecapacity`.")
```
I'm not sure if this suggestion makes total sense but the idea is to note that the cluster itself may have storage available but the node where the storage is attempted to be provisioned does not.

Should this event include a reference to the node where the provisioning was tried? (assuming that information is available at this point in the reconcile) That might also help revise the message a little; something like "Requested storage for is greater than the available storage on `node-1`. Check available capacity by `oc get csistoragecapacity`"

Reply:
Kubernetes checks the capacity in all the nodes and then schedules the Pod referencing this PVC to a Node depending on the available storage. So, there is no node assignment yet. Instead, I'll update the message to have all the capacities of all the nodes: 
```
Requested storage (100000Gi) is greater than available capacity on any node (node ip-10-0-190-168.us-west-1.compute.internal has 2697273016320 bytes of free storage).
```
WDYT?

Reply:
I like it! From a useability point of view it makes it clearer what's being requested and provides evidence that none of the nodes can service the request. One nit, can the node's available space unit match the requested space unit? (e.g. if requested is `Gi` then node space available is also `Gi`)

This will be helpful in the future when we start allowing less than gigabyte sized requests and makes the quick visual comparison easier.

Reply:
I've added a simple function to pretty-print the size: https://github.com/openshift/lvm-operator/compare/f2975147ae757cd8cb27be323b65ae2b14231582..053df2bb1d1d852644faa73d1547b4b94ca50d68